decentraland/unity-renderer
# fix: hiccup opening friends hud

## What does this PR change?
Fix #4680 

We were having a huge hiccup each time we opened the FriendsHUD, specially if we had some friends with a very long list of `muted` or `blocked` users in their profiles.

### Cause
I noticed that the problem was in the way how we were comparing, inside `UserProfileModel.Equals(...)`, if 2 profiles had the same list of `blocked`/`muted` users. We were using:
```
if (!model.blocked.All(blocked.Contains) || model.blocked.Count != blocked.Count)
    return false;
...
if (!model.muted.All(muted.Contains) || model.muted.Count != muted.Count)
    return false;
```
But these `All` operations together with `Contains` to check if a list contains the same elements than another one is very expensive and, in the cases of friends that have a huge amount of blocked and/or muted users set in their profiles (I have seen profiles with more than 10k elements there), was making the loading of the friends list get freezed for several seconds.

### Solution
To find a more efficient way to compare the lists:
```
if (model.blocked.Count != blocked.Count ||
    model.blocked.Except(blocked).Any() ||
    blocked.Except(model.blocked).Any())
    return false;
...
if (model.muted.Count != muted.Count ||
    model.muted.Except(muted).Any() ||
    muted.Except(model.muted).Any())
    return false;
```
#### Edited
We finally decided (read [this](https://github.com/decentraland/unity-renderer/pull/4816#issuecomment-1491662691)) to compare only the lenght of the lists:
```
if (model.blocked.Count != blocked.Count)
    return false;
...
if (model.muted.Count != muted.Count)
    return false;
```

## How to test the changes?
1. Go to: https://play.decentraland.org/?explorer-branch=fix/huge-hiccup-opening-friends-hud
2. Open the FriendsHUD clicking on the corresponding button in the taskbar.
3. Check the list of friends loads quickly just after opening the HUD.

## Our Code Review Standards
https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
